### DISCUSSION TITLE
What is ragna's approach to data connectors?

### DISCUSSION THREAD

**Discussion by nenb** [Quansight/ragna#202] [General]
**Question**: Like the title - how does `ragna` intend to develop support for other data connectors eg different file extensions like `.docx`, or documents that are not on the local filesystem such as in the cloud or behind an API?

**Context:** Similar in spirit to #177. At the moment, `ragna` only has support for reading local files with `.txt` and `.pdf` extensions. This support is provided via classes in the `core/_document.py` module. These classes are built on top of a custom import system that eagerly loads all available file handlers at runtime. The custom import system would seem to suggest that all new data connectors will need to be written from scratch. 

**Personal Thoughts:** On first impressions, this seems to be creating a very large body of work and distracting from `ragna`'s goal as a very accessible and effective RAG orchestration framework eg see the number of data connectors supported by `langchain` [here](https://github.com/langchain-ai/langchain/tree/master/libs/langchain/langchain/document_loaders). Duplicating this effort in `ragna` also seems to me like a questionable use of community resources (but this is a very personal opinion). Is there any way that `ragna` could hook in to the community effort from other projects (but only for the data connectors), and hence allow the `ragna` team to focus on optimising the RAG orchestration experience?

**Comment by pmeier** [comment]
> **Question**: Like the title - how does `ragna` intend to develop support for other data connectors eg different file extensions like `.docx`, or documents that are not on the local filesystem such as in the cloud or behind an API?

These are two different concepts. Let's tackle them individually.

---

> how does `ragna` intend to develop support for other data connectors eg different file extensions like `.docx`

This is a decision that I struggled for a bit when implementing it. Ultimately, I decided to keep it internal for now as part of the bigger refactor in #72. Meaning, for now, if you want to add a new document handler, you need to send a PR to ragna to add it. My reasoning for this is that this is common component in almost all use cases.

If you can show me a good use case where you want a custom `DocumentHandler` that likely won't make its way into Ragna core, I'm happy to reconsider.

> Duplicating this effort in ragna also seems to me like a questionable use of community resources (but this is a very personal opinion). Is there any way that ragna could hook in to the community effort from other projects (but only for the data connectors), and hence allow the ragna team to focus on optimising the RAG orchestration experience?

I soft-disagree here. The code for our `DocumentHandler`s is super simple right now. For example, the whole PDF handling is < 30 LoC:

https://github.com/Quansight/ragna/blob/4962665828c51bc6ef0f2d772574f08caec547a8/ragna/core/_document.py#L237-L263

with the actual core part being only 5

https://github.com/Quansight/ragna/blob/4962665828c51bc6ef0f2d772574f08caec547a8/ragna/core/_document.py#L259-L263

So we aren't looking at crazy amounts of duplication here. IMO it is not worth it to pull a dependency in for this.

That being said, maybe `ragna-langchain` is a good idea for a third-party library that provides adapters between the two projects?

---

> or documents that are not on the local filesystem such as in the cloud or behind an API?

This is what the `Document` class in Ragna is all about. We are defaulting to `LocalDocument`, but there is a (not well documented) example for storing documents in S3: https://github.com/Quansight/ragna/tree/main/examples/s3_documents. Meaning, this is entirely possible already.

**Comment by nenb** [comment]
I recognise that extensions for fairly-arbitrary data connectors are possible at the moment, and I really appreciate the thought that was put in originally to allow this. 

My question though was i) is it the intention for `ragna` to implement support for all data connectors from scratch in the project (based on your answer, this seems to be the case) and ii) whether this is the best use of available resources.

I realise that adding support for a new connector is relatively straightforward (ten's of LoC as you point out). But it is the sheer scale of the connectors that are required (see the `langchain` link in my original message) as well as the maintenance burden that this brings that I was more concerned about. I was hopeful that there might be a way to hook into community support for these connectors in some way (as you point out, they aren't complex or particularly novel), and that would allow focusing more on the orchestration + UX aspects of `ragna`.

**Reply by pmeier** [reply]
> I realise that adding support for a new connector is relatively straightforward (ten's of LoC as you point out). But it is the sheer scale of the connectors that are required (see the langchain link in my original message) as well as the maintenance burden that this brings that I was more concerned about.

I think this is a misunderstanding on your side. We are not trying to replicate the langchain ecosystem in terms of functionality. Ragna is here to provide a well curated set of components that cover the largest chunk of the use cases. For everything else, we have the extension mechanism.

Indeed, the `DocumentHandler`s are currently not available as extensions. But TBH I don't see a point for that right now. We currently have support for .txt and .pdf documents. Throw .doc / .docx and .ppt / .pptx in there (PRs welcome!) we are likely covering 95%+ of the use cases. That is good enough for me until someone actually comes along and has a solid use case where our builtin stuff is not sufficient. Even in that case, we can probably just make `DocumentHandler` a component and let the user extend Ragna that way instead of providing this builtin.

To answer your questions directly:

> i) is it the intention for `ragna` to implement support for all data connectors from scratch in the project (based on your answer, this seems to be the case)

No, it is not the intention to implement everything that langchain has. Yes, we'll currently implement the functionality from scratch for the components we deem useful to have inside Ragna.

> ii) whether this is the best use of available resources.

Yes, I think it is. Using langchain (or any comparable library) as base also has a maintenance burden attached to it. And so far I think this will be higher than implementing a few `DocumentHandler`'s from scratch.

**Reply by nenb** [reply]
> We are not trying to replicate the langchain ecosystem in terms of functionality. Ragna is here to provide a well curated set of components that cover the largest chunk of the use cases. For everything else, we have the extension mechanism.

I'm in agreement with this. On the other hand, RAG tools are only as useful as the knowledge available to them. I think `ragna` should be careful not to make this too restrictive. Limiting the sources of knowledge could place serious limitations on how useful `ragna` can be for many users. 

I've looked a little at the data connectors available to `langchain` (see their `document_loaders`) and `llama_hub`. My assumption is that both of these sources are sufficiently mature that they support the majority of data connectors that most people want (it can't be exhaustive as it's a long tail as you mentioned). Most RAG frameworks (like `quivr` for example) appear to implicitly rely on one of these two sources for data connectors. 

Now let's assume that `ragna` will not hook into either of `langchain` or `llama_index` (I'm personally very sympathetic to this viewpoint if it's possible). My assessment is that the majority of data connectors that are provided by these two packages can be broken into a small number of groups:

**1. human-readable file formats** - these include txt, md, mdx, csv, json, xml, yaml, toml, rst, html and code such as py, js, rs etc.
**2. proprietary file formats** - these include the complete office suite of tools (word for Microsoft, keynote for Apple etc) and pdf
**3. remote 'file systems'** - object storage such as S3 etc, dropbox and github
**4. webpages** - perhaps overlapping with html in 1., but also includes crawling relevant links
**5. external APIs** - slack, gsuite, microsoft suite of tools, notion, confluence, jira, social media such as twitter/linkedin etc, transciption-based services for GMeet/Zoom etc, and probably a bunch more
**6. databases** - SQL and NoSQL

Note that while it's true that all human-readable file formats can be opened with the current `TxtDocumentHandler`, I think the current understanding is that this will not allow you to reach an acceptable level of performance for users. What the majority of RAG tools do at the moment is implement some sort of parser (eg HTML parser, markdown parser etc) for extracting the relevant text. Also note that this is distinct to the chunking and overlap parameters, which are somewhat generic and occur after parsing.

**IMO, what I think `ragna` needs to decide is which of these groups/sub-groups are in scope, and which definitely are not in scope.** I'll add a few possible suggestions that seems sensible to me in the next message.

**Reply by nenb** [reply]
**Local and Remote File Systems**

`ragna` is currently designed for file formats on a local file system with the `LocalDocument` class. Extending this to remote 'file systems' (I'm being lazy and referring to object storage as a file system here) seems like it could be a bit of work, and feels like something that has been re-implemented in a bunch of OSS projects already. What about the use of `fsspec` as an abstraction over these different types of file systems? 

To me, `fsspec` feels like it aligns with `ragna` in terms of lightweight nature and attention to details like serialisability. Perhaps it could be possible to extend the `LocalDocument` class to also include `fsspec` support? This way, it wouldn't matter where a user's pdf file (for example) was located, as long as `ragna` had a pdf handler it would just load it. `ragna` wouldn't have to spend (too much) time implementing support for all the different file systems, and could share the maintenace burden with the rest of the OSS community. What do you think?

**External APIs**

I personally think this is important. I have seen so many use cases for RAG that involve hitting an org's collection of Slack messages, or their archive of decision documents in Google Workspace. I think it would be great to allow some sort of extendability for loading documents from external APIs eg an alternative class to the `LocalDocument` that can be extended by the user. There are so many APIs available, that I don't think that `ragna` could implement them all (see `langchain` for an example of the effort required). Hence why I am suggesting an extension class.

It could be argued that a user should be responsible for exporting data from these external APIs to some file system before using `ragna`. That way, `ragna` wouldn't need to worry about including this support. I'm not sure about this though - some of these tools feel so embedded in daily workflows now, that it would be a shame not to support them. I'd be interested to hear `ragna`s views on this (which can of course change over time too!)

**File formats**

I haven't got any bright ideas for building on community work here. I would personally find it helpful to be explicit about the number of file formats that aim to be supported before starting the work, to get an idea of the amount of work involved. (And I would of course be happy to help implement a bunch of these file format parsers once/if they have been selected!)

**Other**

I'm definitely getting ahead of myself here, but it would be great to try and keep the data connector part of the project as uncoupled as possible to the rest of the project. It would be great to be able to share something like this with the rest of the community, as my impression is that a lightweight (ie minimal depedency and limited LoC) solution doesn't currently exist (and if it does, then perhaps we should consider using it!)

**Reply by pmeier** [reply]
> What about the use of fsspec as an abstraction over these different types of file systems?

Sounds good. Please open a feature request for it. That being said, I would very much prefer [`universal-pathlib`](https://pypi.org/project/universal-pathlib/), which relates to `fsspec` as `os.path` relates to `pathlib`. I never want to deal with str paths again.

> I think it would be great to allow some sort of extendability for loading documents from external APIs eg an alternative class to the LocalDocument that can be extended by the user.

Apart from the fact that I think this should be an external library and not in Ragna core (not saying that the Ragna team cannot maintain something like that), I think the biggest issue is consistency with the UI. Right now we only allow to upload documents from the local machine to some backend. By default this is the machine the API is running on, but can be changed to S3 for example. 

However as soon as you change the intake to no longer being documents from local machine, the UI needs to change as well. And with that comes the big question on how to do that properly? Should we put the relevant UI code on the `Document` class? What happens if the user doesn't even want the UI or wants to use a custom one? Should we still enforce UI code for Ragna to be present on the `Document` class?

Not saying this is impossible, but we need to answer all of these question (and potentially more that I cannot think of right now), before we even start to consider other data sources. This is not impossible, but needs a clean design. This would also bring us closer to a solution for https://github.com/Quansight/ragna/issues/176#issuecomment-1803446728.

> I haven't got any bright ideas for building on community work here. I would personally find it helpful to be explicit about the number of file formats that aim to be supported before starting the work, to get an idea of the amount of work involved. (And I would of course be happy to help implement a bunch of these file format parsers once/if they have been selected!)

IMO, we should to this on demand. So far we only got a request for adding support for Markdown files in #209 (see #210). And I suspect that we will also need support for Word documents and Powerpoint slides in the near future. But other than that, I can't think of any right now. So we shouldn't just add features for the sake of having them. This increases maintenance burden for no reason.

> I'm definitely getting ahead of myself here, but it would be great to try and keep the data connector part of the project as uncoupled as possible to the rest of the project. It would be great to be able to share something like this with the rest of the community, as my impression is that a lightweight (ie minimal depedency and limited LoC) solution doesn't currently exist (and if it does, then perhaps we should consider using it!)

I feel like I'm repeating myself here. Ragna is already really decoupled. You are free to implement any `Document` class that you want to use. If you think this is useful enough for other people as well, throw it in small package and publish it. With your own #192, using these external packages really simple.

IMO, one of the largest mistakes `langchain` made is to have everything inside a single package. Ragna goes a different way. To re-iterate: I'm not saying everything that is not part of Ragna core has to be maintained by the community. Think of it more like `pytest`. They have a super customizable base with clean scope and maintain plenty of plugins themselves: https://github.com/pytest-dev

**Reply by nenb** [reply]
Sorry for taking so long to follow up on this!

> Sounds good. Please open a feature request for it.

I have shared some work in #233. I used `fsspec` rather than `universal-pathlib` largely because it supports more filesystems at present and I am far more familiar with. But I'm open to change here. If the work is of interest, I will then open a feature request. 

>Apart from the fact that I think this should be an external library and not in Ragna core (not saying that the Ragna team cannot maintain something like that), I think the biggest issue is consistency with the UI. 

I'll play around with this over the next couple of weeks, and share anything I find of interest. 

> IMO, we should to this on demand. So far we only got a request for adding support for Markdown files in https://github.com/Quansight/ragna/discussions/209 (see https://github.com/Quansight/ragna/issues/210). And I suspect that we will also need support for Word documents and Powerpoint slides in the near future. But other than that, I can't think of any right now. So we shouldn't just add features for the sake of having them. This increases maintenance burden for no reason.

My two cents: `ragna` is 'only' (I mean this in a very positive way, the minimalism is v attractive and is core design philosophy as you outlined) a RAG orchestrator. Hence, it needs to do RAG well! If I was looking for a RAG tool, and I saw that a package didn't implement the file formats that I needed, then I think I might just move on. Demand might be a bit tricky to measure.

My opinion is that a baseline of .md, .mdx, .pdf, HTML, .txt, a range of office-suite tools (.doc, .docx, .ppt, .pptx at least), and also .csv + .json is necessary. Would you agree? If so, I will open issues for the formats that aren't currently supported, and will start adding support over the next few weeks.

**Reply by pmeier** [reply]
> My opinion is that a baseline of .md, .mdx, .pdf, HTML, .txt, a range of office-suite tools (.doc, .docx, .ppt, .pptx at least), and also .csv + .json is necessary. Would you agree? If so, I will open issues for the formats that aren't currently supported, and will start adding support over the next few weeks.

- .md, .mdx :+1: see #210
- .pdf :+1: already supported
- HTML :shrug: not against it, but we should discuss the proper use case first. Let's do so on the issue.
- .txt :+1: already supported
- .doc, .docx :+1: see #225
- .ppt, .pptx :+1:
- .csv :-1: CSV is usually for structured numerical data, which is kinda on the opposite of the spectrum of a text document. Unless there is a convincing use case that I'm missing, I'm against this. If you feel strongly, please open an issue and detail why Ragna should support this
- .json :shrug: My gut says that JSON is too variable for us to safely extract relevant text. I would need to see an actual use case to make a call here. On the other hand, you are not the first person to request that. See https://github.com/Quansight/ragna/issues/224#issuecomment-1838864057

**Reply by peachkeel** [reply]
I recant my comment in #224 about JSON. The way we're using JSON does not really fit the document handler paradigm. I might start a separate discussion about my use-case,  which has more to do with corpora than documents.
